### ISSUE TITLE
@client query depends on server data

### ISSUE THREAD

**Issue by MattReimer** [apollographql/apollo-link-state#307]
I have a case where my derived state depends on a query to the server. I realize that `cache.readQuery` only reads from the cache so if I haven't already run the `projQuery` I won't get anything back.

Sure enough I get the dreaded error: 

```
Error: Can't find field Project({"ProjectId":1}) on object (ROOT_QUERY) {`
```

So how can I write my resolver with a dependency to the server data? All I have t make queries here is the `cache` object and that doesn't seem to have a `query` function attached to it?

My resolver that I use with `withClientState` looks like this:

```javascript
import gql from 'graphql-tag'

// This is the query on which I depend
const projQuery = gql`
  query getProjectTaxes($ProjectId: Int!) {
    Project(ProjectId: $ProjectId) {
      ProjectId
      Taxonomies {
        TaxonomyId
        Meta
      }        
    }  
  }`

const resolvers = {
  Query: {
    ProjectDerived: (_obj, {ProjectId}, {cache}, info) => {
      const projQueryRes = cache.readQuery({
        query: projQuery,
        variables: { ProjectId }
      })
      /* SOME logic here using the result of projQueryRes */
      const resultOfLogic = true

      return { myVariable: resultOfLogic }
    }
  }
}
export default resolvers

```

**Comment by hwillson** [comment]
@MattReimer We're working on a new version of `apollo-link-state` that replaces the concept of `defaults` with more dynamic function based `initializers`. These initializer functions can be run to prime the cache in whatever way you see fit, and have access to both the cache and the instantiated Apollo Client instance. These new initializer functions can also be run at any point; they're not limited to being called when Apollo Client is instantiated or when your link chain is first constructed. To address your issue here, you could create a new initializer function that takes care of priming the cache the way you need, by running (async) your initial network query. So in other words, you can do something like:

```js
client.addLocalStateInitializers({
  async projectTaxes(cache, client) {
    // ... run your query, prime the cache, etc. 
  },
});
```

Would this work for you?

**Comment by MattReimer** [comment]
Yeah! I think this would be fantastic. Sounds exactly like what I need.

**Comment by arif-hanif** [comment]
@hwillson that makes sense.

**Comment by Borales** [comment]
@hwillson Hi! I've seen your "_we're working on a new version of `apollo-link-state`_" response on different issues for this repo, do you have any preview for these changes and estimated date of new release?

**Comment by smolinari** [comment]
*me wants to know too*

Scott

**Comment by pabacham** [comment]
@hwillson any news about release date?

**Comment by tsaiDavid** [comment]
@hwillson - would love to test out the beta, ran into this problem with our product

**Comment by smolinari** [comment]
It's November 7th. Please do give us an update about the update. :)

Scott

**Comment by Borales** [comment]
Looks like `apollo-link-state` is being merged to `apollo-client` (https://github.com/apollographql/apollo-client/compare/hwillson/link-state-merge)

**Comment by smolinari** [comment]
Ohh.. that is a lot of opinionation to be packing into the core. 

Scott

**Comment by Borales** [comment]
https://summit-2018-apollo-client.netlify.com/ "**What's next for Apollo Client**" presentation slides from yesterday (there are a few words about state management - it's going to be merge into `apollo-client`)

**Comment by pulpfree** [comment]
Any news on this, ran into same problem looking for solution, even something quick and dirty to solve problem. Is there a way to use ‘client’ in resolver to make query?

**Comment by simonedavico** [comment]
@pulpfree the client is available in the `context` parameter of the resolver. I am also facing the same use case and wonder if this is the right way to go...
